[feat] Add backend infrastructure and proto for audio in ChatInput

sfc-gh-nbellante opened the pull request:
## Describe your changes

Added python support for audio files in the chat input component. 

Key changes:

- Added `audio` field to `ChatInputValue` class to store a single audio file
- Implemented `_pop_audio_file` function to extract audio files from protobuf messages
- Updated serialization/deserialization logic to handle audio files
- Added `accept_audio` flag to the ChatInput proto
- Updated type annotations to accommodate the new audio file field
- Added comprehensive tests for the audio file functionality

## Testing Plan

- Unit Tests: Added Python tests to verify the audio file handling in `ChatInputValue`, including:
    - Testing that audio files are properly handled and accessible
    - Verifying the behavior when audio is None
    - Testing the dict-like interface with the new audio field
    - Confirming that the to_dict method properly includes the audio field

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
Add backend infrastructure and proto for audio in ChatInput

This PR adds the foundational backend plumbing and protocol definitions
to support audio recording in chat_input, without exposing the feature
to users yet.

**Backend changes:**
- Add `audio` field to `ChatInputValue` type to hold recorded audio files
- Add `_pop_audio_file()` helper to extract audio from proto messages
- Update `ChatInputSerde.deserialize()` to handle audio file uploads
- Add comprehensive Python unit tests for audio handling

**Proto changes:**
- Add `ChatInputValue.audio_file_info` field (Common.proto)
- Add `ChatInput.accept_audio` field (ChatInput.proto)

The feature remains dark-shipped: the Python API parameter `accept_audio`
is not included in this PR and will be added separately. The protocol and
backend can handle audio, but users cannot enable it yet.

Testing: New Python unit tests validate audio file handling in
ChatInputValue and deserialization logic.

sfc-gh-nbellante commented:
* **#12833** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12833?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 2  dependent PRs ([#12834](https://github.com/streamlit/streamlit/pull/12834) <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12834?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>, [#12835](https://github.com/streamlit/streamlit/pull/12835) <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12835?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#12832** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12832?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12832?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-nbellante commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -197,6 +199,40 @@ def _pop_upload_files(
     return collected_files
 
 
+def _pop_audio_file(
```
suggestion: let's have this function validate that it was indeed a wav file that was sent up.

sfc-gh-nbellante commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -197,6 +199,40 @@ def _pop_upload_files(
+    file_rec = file_recs_list[0]
+    uploaded_file = UploadedFile(file_rec, audio_file_info.file_urls)
+
+    if hasattr(ctx.uploaded_file_mgr, "remove_file"):
```
concern: This seems fragile. Why is this needed? Why can't we be certain that remove_file exists?

sfc-gh-nbellante commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -604,6 +646,11 @@ def chat_input(
             https://doc-chat-input-session-state.streamlit.app/
             height: 350px
         """
+        if accept_file not in {True, False, "multiple", "directory"}:
```
concern: this seems like an unnecessary change to move this validation up. Why was it moved up?

sfc-gh-nbellante commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -694,7 +736,7 @@ def chat_input(
             ctx=ctx,
             value_type="chat_input_value",
-        )
+        )  # type: ignore[misc]
```
nit: why was this comment moved?

sfc-gh-nbellante committed:
Add WAV file validation to chat input audio files

Validates both file extension (.wav) and MIME type (audio/wav, audio/wave,
audio/x-wav) when processing audio files in chat input. Raises clear
StreamlitAPIException if validation fails to provide better error messages
and prevent non-WAV files from being processed.

Addresses PR feedback comment #1.

sfc-gh-nbellante committed:
Document hasattr check for remove_file method

Added explanatory comment for the hasattr check when calling remove_file.
The check exists because remove_file is not part of the UploadedFileManager
Protocol definition, but is implemented by MemoryUploadedFileManager. This
maintains compatibility with potential alternative implementations.

Addresses PR feedback comment #2.

sfc-gh-nbellante committed:
Revert accept_file validation to original location

Moved the accept_file validation back to its original position (after
check_widget_policies) to keep the PR diff focused on the audio input
feature. This removes unnecessary code churn that makes review harder.

Addresses PR feedback comment #3.

sfc-gh-nbellante committed:
Revert type ignore comment to original location

Moved the type: ignore[misc] comment back to the register_widget( line
instead of the closing paren. This was an unintentional move that added
noise to the diff.

Addresses PR feedback comment #4.

sfc-gh-nbellante committed:
Improve _pop_audio_file documentation and defensive programming

- Add Numpydoc-style docstring with Parameters and Returns sections
- Add explicit audio_file_info check to file removal guard for safety

Addresses Graphite Agent feedback on PR #12832

sfc-gh-nbellante commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -197,6 +199,70 @@ def _pop_upload_files(
+    file_rec = file_recs_list[0]
+    uploaded_file = UploadedFile(file_rec, audio_file_info.file_urls)
+
+    # Validate that the file is a WAV file by checking extension and MIME type
+    if not uploaded_file.name.lower().endswith(".wav"):
+        raise StreamlitAPIException(
+            f"Invalid file extension for audio input: `{uploaded_file.name}`. "
+            "Only WAV files (.wav) are accepted."
+        )
```
Good catch! I'll extract these into module-level constants to improve maintainability and make it easier to support additional audio formats in the future.

sfc-gh-nbellante committed:
Address PR feedback: WAV Validation Constants

Extract magic strings for audio file validation into module-level constants
(_ACCEPTED_AUDIO_EXTENSION and _ACCEPTED_AUDIO_MIME_TYPES) to improve
maintainability and make it easier to support additional audio formats.

Resolves comment 1

sfc-gh-nbellante commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -74,14 +75,15 @@
 class ChatInputValue(MutableMapping[str, Any]):
```
I'll add the Numpydoc-style docstring you suggested.

sfc-gh-nbellante committed:
Address PR feedback: ChatInputValue Docstring

Add comprehensive Numpydoc-style docstring to ChatInputValue dataclass,
documenting its purpose, attributes (text, files, audio), and dict-like
interface for users of the public API.

Resolves comment 2

sfc-gh-nbellante committed:
Address PR feedback: Raises Documentation

Add "Raises" section to _pop_audio_file docstring to document the
StreamlitAPIException that can be raised during audio file validation,
following Numpydoc conventions.

Resolves comment 3

sfc-gh-nbellante commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -197,6 +199,70 @@ def _pop_upload_files(
+    # Note: remove_file is not part of the UploadedFileManager Protocol, but is
+    # implemented by MemoryUploadedFileManager. We use hasattr to maintain
+    # compatibility with potential alternative implementations.
+    if audio_file_info and hasattr(ctx.uploaded_file_mgr, "remove_file"):
```
Going to keep it in here because it's a cheap check and can help in future refactors.

sfc-gh-lwilby commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -197,6 +231,75 @@ def _pop_upload_files(
+
+    # Remove the file from the manager after creating the UploadedFile object.
+    # Note: remove_file is not part of the UploadedFileManager Protocol, but is
+    # implemented by MemoryUploadedFileManager. We use hasattr to maintain
```
[suggestion]: Would it make sense to use `type()` or `isInstance` here? That way, if we start using a different type of `UploadedFileManager` we would need to explicitly add it here. The new class may implement `remove_file`, but we might not want to use it here for that class.

sfc-gh-nbellante replied:
Good call, I've updated it as suggested!

sfc-gh-nbellante committed:
Address PR feedback: hasattr vs isinstance Type Check

Replace hasattr() checks with isinstance(MemoryUploadedFileManager)
for explicit type safety, ensuring only approved manager types use
the cleanup logic.

Resolves comment 1

sfc-gh-nbellante commented:
### Merge activity

* **Nov 1, 7:02 AM UTC**: @sfc-gh-nbellante merged this pull request with [Graphite](https://app.graphite.dev/github/pr/streamlit/streamlit/12832).
